File: Array_Functions/Permutations/permutations_iterative.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

template <typename T>
vector<vector<T>> generatePermutations(vector<T> arr) {
    vector<vector<T>> permutations;
    sort(arr.begin(), arr.end());
    permutations.push_back(arr);
    while (true) {
        int i = arr.size()-2;
        while (i >= 0 && arr[i] >= arr[i + 1]) {
            i--;
        }
        if (i < 0) break;
        int j = arr.size()-1;
        while (arr[j] <= arr[i]) {
            j--;
        }
        swap(arr[i], arr[j]);
        reverse(arr.begin() + i + 1, arr.end());
        permutations.push_back(arr);
    }
    return permutations;
}

int main() {
    vector<int> arr = {1, 3, 2};
    vector<vector<int>> permutations = generatePermutations(arr);

    for (const auto& perm : permutations) {
        for (const auto& elem : perm) {
            cout << elem << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Array_Functions/Permutations/permutations_iterative.py
def generate_permutations(arr):
    permutations = []
    arr.sort()
    permutations.append(arr[:])

    while True:
        i = len(arr) - 2
        while i >= 0 and arr[i] >= arr[i + 1]:
            i -= 1
        if i < 0:
            break
        j = len(arr) - 1
        while arr[j] <= arr[i]:
            j -= 1

        arr[i], arr[j] = arr[j], arr[i]
        arr = arr[:i + 1] + arr[i + 1:][::-1]
        permutations.append(arr[:])

    return permutations

if __name__ == "__main__":
    arr = ['a', 'b', 'c', 'd']
    permutations = generate_permutations(arr)

    for perm in permutations:
        print(" ".join(map(str, perm)))
